app/src/main/java/tech/relaycorp/gateway/domain/publicsync/RegisterGateway.kt

```diff
@@ -0,0 +1,50 @@
+package tech.relaycorp.gateway.domain.publicsync
+
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.data.model.RegistrationState
+import tech.relaycorp.gateway.data.preference.PublicGatewayPreferences
+import tech.relaycorp.gateway.domain.LocalConfig
+import tech.relaycorp.poweb.PoWebClient
+import tech.relaycorp.relaynet.bindings.pdc.ClientBindingException
+import tech.relaycorp.relaynet.bindings.pdc.ServerException
+import tech.relaycorp.relaynet.messages.control.PrivateNodeRegistration
+import java.net.URL
+import java.util.logging.Level
+import javax.inject.Inject
+
+@JvmSuppressWildcards
+class RegisterGateway
+@Inject constructor(
+    private val publicGatewayPreferences: PublicGatewayPreferences,
+    private val localConfig: LocalConfig,
+    private val poWebClientBuilder: ((String) -> PoWebClient)
+) {
+
+    suspend fun registerIfNeeded() {
+        if (publicGatewayPreferences.getRegistrationState() != RegistrationState.ToDo) {
+            return
+        }
+
+        val pnr = register() ?: return
+        localConfig.setCertificate(pnr.privateNodeCertificate)
+        publicGatewayPreferences.setCertificate(pnr.gatewayCertificate)
+        publicGatewayPreferences.setRegistrationState(RegistrationState.Done)
+    }
+
+    private suspend fun register(): PrivateNodeRegistration? =
+        try {
+            val address = publicGatewayPreferences.getAddress()
+            val hostName = URL(address).host
+            val poWeb = poWebClientBuilder.invoke(hostName)
+            val keyPair = localConfig.getKeyPair()
+
+            val pnrr = poWeb.preRegisterNode(keyPair.public)
+            poWeb.registerNode(pnrr.serialize(keyPair.private))
```

Comment:
Can you please wrap the remote method calls in a `poweb.use { ... }` block?

Reply:
Also, I don't want to cause a merge conflict with the stuff you'll change, so could you please log the address of the private gateway when registration succeeds? That'd make it easier for me to find it on the server logs. E.g.,

```
logger.log(Level.INFO, "Successfully registered with address ${registration.privateNodeCertificate.subjectPrivateAddress}")
```